evm/src/trace/identifier/signatures.rs

```diff
@@ -0,0 +1,135 @@
+                serde_json::from_reader(std::fs::File::open(&path)?)?
+            } else {
+                if let Err(err) = std::fs::create_dir_all(cache_path) {
+                    warn!(target: "signaturesidentifier", "could not create signatures cache dir: {:?}", err);
+                }
+                CachedSignatures::default()
+            };
+            return Ok(Self { cached, cached_path: Some(path) })
+        }
+        Ok(Self::default())
+    }
+
+    pub fn save(&self) {
+        if let Some(cached_path) = &self.cached_path {
+            if let Ok(file) = std::fs::File::create(&cached_path) {
+                if serde_json::to_writer(BufWriter::new(file), &self.cached).is_err() {
+                    warn!(target: "signaturesidentifier", "could not serialize SignaturesIdentifier");
+                }
+            } else {
+                warn!(target: "signaturesidentifier", "could not open {}", cached_path.to_string_lossy());
+            }
+        }
+    }
+}
+
+impl SignaturesIdentifier {
+    async fn identify<T>(
+        &mut self,
+        selector_type: SelectorType,
+        identifier: &[u8],
+        get_type: fn(&str) -> eyre::Result<T>,
+    ) -> Option<T> {
+        let map = match selector_type {
+            SelectorType::Function => &mut self.cached.functions,
+            SelectorType::Event => &mut self.cached.events,
+        };
+
+        let identifier = format!("0x{}", hex::encode(identifier));
+
+        if !map.contains_key(&identifier) {
+            if let Ok(mut signatures) = decode_selector(&identifier, selector_type).await {
+                if !signatures.is_empty() {
+                    map.insert(identifier.to_string(), signatures.pop().expect(""));
+                }
+            }
+        }
+
+        if let Some(signature) = map.get(&identifier) {
+            return get_type(signature).ok()
+        }
+        None
+    }
+
+    /// Identifies `Function` from its cache or `sig.eth.samczsun.com`
+    pub async fn identify_function(&mut self, identifier: &[u8]) -> Option<Function> {
+        self.identify(SelectorType::Function, identifier, get_func).await.map(|mut function| {
+            function.name = format!("*{}", function.name);
+            function
+        })
+    }
+
+    /// Identifies `Event` from its cache or `sig.eth.samczsun.com`
+    pub async fn identify_event(&mut self, identifier: &[u8]) -> Option<Event> {
+        self.identify(SelectorType::Event, identifier, get_event).await
+    }
+}
+
+impl Drop for SignaturesIdentifier {
+    fn drop(&mut self) {
+        self.save();
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize, Default)]
+pub struct CachedSignatures {
+    pub events: BTreeMap<String, String>,
+    pub functions: BTreeMap<String, String>,
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[tokio::test]
+    async fn can_query_signatures() {
+        let _ = std::fs::remove_file("./signatures");
+        {
```

Comment:
should we be using a tmpfile here instead? so that we don't end up having lingering files